### PR TITLE
Feat/met 120 code refactoring for the enrichment and dereferencing services

### PR DESCRIPTION
**Pull request by jochen-vermeulen**

### CODE REVIEW

File: metis-dereference/metis-dereference-service/src/test/java/eu/europeana/metis/derefrence/test/MongoDereferenceServiceTest.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 import eu.europeana.enrichment.api.external.EntityWrapper;
```
Comment (jochen-vermeulen) [inline metis-dereference/metis-dereference-service/src/test/java/eu/europeana/metis/derefrence/test/MongoDereferenceServiceTest.java]: Is there a reason for all the System.out.println statements? They don't really belong in test code.

**Comment by jochen-vermeulen** [comment]
What role has the file cclient.properties? I don't think it should be there?

File: metis-dereference/metis-dereference-client/src/test/resources/client.properties.example [hunk 2]
```diff
@@ -0,0 +1 @@
+host.url=
```
Comment (jochen-vermeulen) [inline metis-dereference/metis-dereference-client/src/test/resources/client.properties.example]: You copied this file to cclient.properties in the main resource folder. I don't think cclient.properties should be there, this file is sufficient I think.

File: metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java [hunk 3]
```diff
@@ -26,17 +38,8 @@
...
-import org.apache.commons.io.IOUtils;
-import org.apache.commons.lang3.StringUtils;
-import org.jibx.runtime.BindingDirectory;
-import org.jibx.runtime.IBindingFactory;
-import org.jibx.runtime.IUnmarshallingContext;
-import org.jibx.runtime.JiBXException;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
+import eu.europeana.enrichment.api.external.model.EnrichmentBase;
+import eu.europeana.enrichment.utils.EntityMergeUtils;
```
Comment (jochen-vermeulen) [inline metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java]: In this class there is a System.exit(-1). That may have unexpected consequences if other applications start using it... maybe we should change this?

Comment (jochen-vermeulen) [inline metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java]: This class has quite a few SonarQube issues. Maybe we can try to eliminate some of them?

File: metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java [hunk 4]
```diff
@@ -68,29 +71,58 @@ private DereferenceUtils() { }
...
+	    values.addAll(dereferenceTimespanList(rdf.getTimeSpanList()));
+	    values.addAll(dereferenceWebResourceList(rdf.getWebResourceList()));
+	    values.addAll(dereferenceProxyList(rdf.getProxyList()));
 
-    return values;
+	    return values;
   }
+  
+  public static RDF toRDF(String xml) throws JiBXException {
+	    Set<String> values = new HashSet<>();
```
Comment (jochen-vermeulen) [inline metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java]: This value is never used!

File: metis-dereference/metis-dereference-common/src/test/java/eu/europeana/metis/dereference/DereferenceUtilsTest.java [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+        Assert.assertTrue(result.contains("http://dummy27.dum"));
+        Assert.assertTrue(result.contains("http://dummy28.dum"));
+        Assert.assertTrue(result.contains("http://dummy29.dum"));
+        Assert.assertTrue(result.contains("http://dummy30.dum"));
+        
+        Assert.assertTrue(result.contains("http://dummy31.dum"));
+        Assert.assertTrue(result.contains("http://dummy32.dum"));
+        Assert.assertTrue(result.contains("http://dummy33.dum"));
+        Assert.assertTrue(result.contains("http://dummy34.dum"));
+        Assert.assertTrue(result.contains("http://dummy35.dum"));
```
Comment (jochen-vermeulen) [inline metis-dereference/metis-dereference-common/src/test/java/eu/europeana/metis/dereference/DereferenceUtilsTest.java]: If you automate the creation of these dummy URLS you should be able to get rid of these long assert blocks but instead put them in a for loop. This might make it much more readable. The same goes for other methods.

File: metis-enrichment/metis-enrichment-common/src/test/java/eu/europeana/enrichment/utils/EnrichmentUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+		 
+		 // Should be rejected
+		 Choice choice11 = new Choice();
+		 Extent extent = new Extent();
+		 extent.setString("Extent");
+		 Lang language11 = new Lang();
+		 language11.setLang("Ukrainian");
+		 extent.setLang(language11);
+		 choice11.setExtent(extent);
+		 choiceList.add(choice11);
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/test/java/eu/europeana/enrichment/utils/EnrichmentUtilsTest.java]: Maybe put all the values of these choices (like language) in constant arrays so that you can loop over the creation and assertion of these choices?

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java [hunk 7]
```diff
@@ -21,20 +24,22 @@
  */
 public class EnrichmentUtils {
     public static final Logger LOGGER = LoggerFactory.getLogger(EnrichmentUtils.class);
-
-    private static IBindingFactory rdfFactory;
+    private static IBindingFactory rdfFactory, placeTypeFactory;
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java]: placeTypeFactory is not used.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java [hunk 8]
```diff
@@ -21,20 +24,22 @@
  */
 public class EnrichmentUtils {
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java]: There are two SonarQube issues in this class. They are easy to solve.

Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java]: Some methods in this class are not covered by unit tests. These methods are fairly straightforward, and typically wrap around another method/library. But maybe there is something we want to check? How to deal with edge cases and what exceptions to expect, and so on?

File: metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java (hunk 3 above)
Comment (jochen-vermeulen) [inline metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java]: Some methods in this class are not covered by unit tests. These methods are fairly straightforward, and typically wrap around another method/library. But maybe there is something we want to check? How to deal with edge cases and what exceptions to expect, and so on?

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java [hunk 9]
```diff
@@ -0,0 +1,899 @@
...
+import org.jibx.runtime.IBindingFactory;
+import org.jibx.runtime.IUnmarshallingContext;
+import org.jibx.runtime.JiBXException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Created by erikkonijnenburg on 28/07/2017.
+ */
+public class EntityMergeUtils {
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java]: This class is quite important but has no unit tests. It might be desirable to add some.

Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java]: There are tons of SonarQube issues with this class. I think it would be good to fix them as this class is quite essential to what we are trying to do here.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java [hunk 10]
```diff
@@ -0,0 +1,899 @@
...
+  static {
+    try {
+      agentFactory = BindingDirectory.getFactory(AgentType.class);
+      conceptFactory = BindingDirectory.getFactory(Concept.class);
+      placeFactory = BindingDirectory.getFactory(PlaceType.class);
+      tsFactory = BindingDirectory.getFactory(TimeSpanType.class);
+      rdfFactory = BindingDirectory.getFactory(RDF.class);
+    } catch (JiBXException e) {
+      LOGGER.error("Unable to get BindingFactory", e);
+      System.exit(-1);
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java]: This seems a bit dramatic and dangerous (this class is to be used as a library, a hard exit can have unexpected consequences). Maybe we are better off with some other method of checking for errors.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/InputValue.java [hunk 11]
```diff
@@ -14,7 +14,7 @@
  *  See the Licence for the specific language governing permissions and limitations under
  *  the Licence.
  */
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/InputValue.java]: This class will be part of the API of the library. If we want to be extra safe, we could make this class unmodifiable (give it a constructor, remove all setters and copy/clone all arrays). What do you think?

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/InputValue.java [hunk 12]
```diff
@@ -14,7 +14,7 @@
...
-package eu.europeana.metis.utils;
+package eu.europeana.enrichment.utils;
 
 import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 import com.fasterxml.jackson.databind.annotation.JsonSerialize.Inclusion;
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/InputValue.java]: Can we get rid if this deprecated reference?

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityClass.java [hunk 13]
```diff
@@ -14,7 +14,7 @@
  *  See the Licence for the specific language governing permissions and limitations under
  *  the Licence.
  */
-package eu.europeana.metis.utils;
+package eu.europeana.enrichment.utils;
 
 import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 import com.fasterxml.jackson.databind.annotation.JsonSerialize.Inclusion;
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityClass.java]: Can we get rid of this deprecated reference?

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java [hunk 14]
```diff
@@ -1,4 +1,4 @@
-package eu.europeana.metis.utils;
+package eu.europeana.enrichment.utils;
 
 import eu.europeana.corelib.definitions.jibx.Contributor;
 import eu.europeana.corelib.definitions.jibx.Coverage;
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java]: There is a lot of duplicate code in this class. There must be a way to make this more generic.

Comment (jochen-vermeulen) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java]: This class has easy to solve SonarQube issues.

Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java]: This class has easy to solve SonarQube issues.

Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java]: This class could probably do with some JavaDoc.

File: metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/EnrichmentWorker.java [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+import eu.europeana.enrichment.utils.InputValue;
+import eu.europeana.metis.dereference.DereferenceUtils;
+import eu.europeana.metis.dereference.client.DereferenceClient;
+
+/*
+ * To be integrated into eCloud topology. This code serves as an example of enrichment through use of the
+ * enrichment and dereferencing clients and utils. 
+ * 
+ */
+public class EnrichmentWorker {
```
Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/EnrichmentWorker.java]: This class has SonarQube issues (beyond the use of System.out.println) that should be solved.

Comment (jochen-vermeulen) [inline metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/EnrichmentWorker.java]: Maybe we can add more comments to this class? We are trying to tell them how to use the code that we supply. Maybe it would be great for them to have some explanation on what the calls do.

Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/EnrichmentWorker.java]: Some extra comments have been added but overall the flow of the logic (steps), together with the related usage of the utils/clients is quite clear.

Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-client/src/main/java/eu/europeana/enrichment/rest/client/EnrichmentWorker.java]: Not worth the time investment. This class is just a placeholder/container for the logic that will be implemented in the eCloud topology.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java (hunk 14 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java]: This entire class needs a massive rewrite/refactor. This can best be done in upcoming sprints after we come up with a larger cleanup strategy together.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityClass.java (hunk 13 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityClass.java]: Done.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/InputValue.java (hunk 12 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/InputValue.java]: Done.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/InputValue.java (hunk 11 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/InputValue.java]: This should be part of our larger cleanup strategy.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java (hunk 14 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java]: See previous comment about this class.

Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentFields.java]: See previous comment about this class.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java (hunk 10 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java]: The System.exit is called in the static initializer block and as such I don't see a problem with this. Since this is a crucial class, the classloader should be able to load it without an error/exception and if there is an error, stopping is not a crazy option. This static block should never be catching an exception and then letting the application continue.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java (hunk 9 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java]: Agreed, but as this class is never called directly at this point (only via the enrichment and deference utils classes which do have unit test coverage), we can probably let this slide for now.

File: metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java (hunk 3 above)
Comment (SidSatbhai) [reply metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java]: The methods used for the code that will be used in the enrichment topology is covered by unit tests. Further coverage should come after we decide if we're going to refactor/modify things significantly.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java (hunk 8 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java]: The methods used for the code that will be used in the enrichment topology is covered by unit tests. Further coverage should come after we decide if we're going to refactor/modify things significantly.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java (hunk 7 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java]: Removed.

File: metis-enrichment/metis-enrichment-common/src/test/java/eu/europeana/enrichment/utils/EnrichmentUtilsTest.java (hunk 6 above)
Comment (SidSatbhai) [reply metis-enrichment/metis-enrichment-common/src/test/java/eu/europeana/enrichment/utils/EnrichmentUtilsTest.java]: That would look neater, yes. But, later on, if you want to make more specific test cases with fringe choice configurations, you'll have to write code like this anyway.

File: metis-dereference/metis-dereference-common/src/test/java/eu/europeana/metis/dereference/DereferenceUtilsTest.java (hunk 5 above)
Comment (SidSatbhai) [reply metis-dereference/metis-dereference-common/src/test/java/eu/europeana/metis/dereference/DereferenceUtilsTest.java]: Later on, more complex assertions might be required that use more than just the value of the resource field for the various objects created.

File: metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java (hunk 4 above)
Comment (SidSatbhai) [reply metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java]: Removed.

File: metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java (hunk 3 above)
Comment (SidSatbhai) [reply metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java]: Again, it's in the static block, so I don't really mind this.

File: metis-dereference/metis-dereference-client/src/test/resources/client.properties.example (hunk 2 above)
Comment (SidSatbhai) [reply metis-dereference/metis-dereference-client/src/test/resources/client.properties.example]: This file has been removed.

File: metis-dereference/metis-dereference-service/src/test/java/eu/europeana/metis/derefrence/test/MongoDereferenceServiceTest.java (hunk 1 above)
Comment (SidSatbhai) [reply metis-dereference/metis-dereference-service/src/test/java/eu/europeana/metis/derefrence/test/MongoDereferenceServiceTest.java]: Agreed. I've removed all those pesky printlns.

File: metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java (hunk 3 above)
Comment (jochen-vermeulen) [reply metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java]: I fixed it.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java (hunk 8 above)
Comment (jochen-vermeulen) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EnrichmentUtils.java]: Solved.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java (hunk 10 above)
Comment (jochen-vermeulen) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java]: I fixed it.

File: metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java (hunk 3 above)
Comment (jochen-vermeulen) [reply metis-dereference/metis-dereference-common/src/main/java/eu/europeana/metis/dereference/DereferenceUtils.java]: I fixed some of them.

File: metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java (hunk 9 above)
Comment (jochen-vermeulen) [reply metis-enrichment/metis-enrichment-common/src/main/java/eu/europeana/enrichment/utils/EntityMergeUtils.java]: I fixed them.
